src/event.rs

```diff
@@ -0,0 +1,619 @@
+	pub(crate) fn event_handled(&self) -> Result<(), Error> {
+		{
+			let mut locked_queue = self.queue.lock().unwrap();
+			locked_queue.pop_front();
+			self.persist_queue(&*locked_queue)?;
+		}
+		self.notifier.notify_one();
+		Ok(())
+	}
+
+	fn persist_queue(&self, locked_queue: &VecDeque<Event>) -> Result<(), Error> {
+		self.persister
+			.persist(EVENTS_PERSISTENCE_KEY, &EventQueueSerWrapper(locked_queue))
+			.map_err(|_| Error::PersistenceFailed)?;
+		Ok(())
+	}
+}
+
+impl<K: Deref> ReadableArgs<K> for EventQueue<K>
+where
+	K::Target: KVStorePersister,
+{
+	#[inline]
+	fn read<R: lightning::io::Read>(
+		reader: &mut R, persister: K,
+	) -> Result<Self, lightning::ln::msgs::DecodeError> {
+		let read_queue: EventQueueDeserWrapper = Readable::read(reader)?;
+		let queue: Mutex<VecDeque<Event>> = Mutex::new(read_queue.0);
+		let notifier = Condvar::new();
+		Ok(Self { queue, notifier, persister })
+	}
+}
+
+struct EventQueueDeserWrapper(VecDeque<Event>);
+
+impl Readable for EventQueueDeserWrapper {
+	fn read<R: lightning::io::Read>(
+		reader: &mut R,
+	) -> Result<Self, lightning::ln::msgs::DecodeError> {
+		let len: u16 = Readable::read(reader)?;
+		let mut queue = VecDeque::with_capacity(len as usize);
+		for _ in 0..len {
+			queue.push_back(Readable::read(reader)?);
+		}
+		Ok(Self(queue))
+	}
+}
+
+struct EventQueueSerWrapper<'a>(&'a VecDeque<Event>);
+
+impl Writeable for EventQueueSerWrapper<'_> {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), lightning::io::Error> {
+		(self.0.len() as u16).write(writer)?;
+		for e in self.0.iter() {
+			e.write(writer)?;
+		}
+		Ok(())
+	}
+}
+
+pub(crate) struct EventHandler<K: Deref, L: Deref>
+where
+	K::Target: KVStorePersister,
+	L::Target: Logger,
+{
+	wallet: Arc<Wallet<bdk::sled::Tree>>,
+	event_queue: Arc<EventQueue<K>>,
+	channel_manager: Arc<ChannelManager>,
+	network_graph: Arc<NetworkGraph>,
+	keys_manager: Arc<KeysManager>,
+	inbound_payments: Arc<PaymentInfoStorage>,
+	outbound_payments: Arc<PaymentInfoStorage>,
+	tokio_runtime: RwLock<Option<Arc<tokio::runtime::Runtime>>>,
+	logger: L,
+	_config: Arc<Config>,
+}
+
+impl<K: Deref, L: Deref> EventHandler<K, L>
+where
+	K::Target: KVStorePersister,
+	L::Target: Logger,
+{
+	pub fn new(
+		wallet: Arc<Wallet<bdk::sled::Tree>>, event_queue: Arc<EventQueue<K>>,
+		channel_manager: Arc<ChannelManager>, network_graph: Arc<NetworkGraph>,
+		keys_manager: Arc<KeysManager>, inbound_payments: Arc<PaymentInfoStorage>,
+		outbound_payments: Arc<PaymentInfoStorage>, logger: L, _config: Arc<Config>,
+	) -> Self {
+		let tokio_runtime = RwLock::new(None);
+		Self {
+			event_queue,
+			wallet,
+			channel_manager,
+			network_graph,
+			keys_manager,
+			inbound_payments,
+			outbound_payments,
+			logger,
+			tokio_runtime,
+			_config,
+		}
+	}
+
+	pub(crate) fn set_runtime(&self, tokio_runtime: Arc<tokio::runtime::Runtime>) {
+		*self.tokio_runtime.write().unwrap() = Some(tokio_runtime);
+	}
+
+	pub(crate) fn drop_runtime(&self) {
```

Comment:
No, currently we don't drop *all* LDK objects on `stop()`, but just the runtime-dependent ones. So in order to make sure the runtime itself is dropped, we also need to drop the references here.

Reply:
The current idea is to keep the LDK objects in memory, and only stop/resume operation via these methods.

Reply:
I'm not seeing the `EventHandler` or `BackgroundProcessor` used there. I was thinking that given `BackgroundProcessor::start` (and `process_events_async`) takes ownership of `EventHandler` and doesn't need to be persisted, then it can dropped on `stop()`. And since `EventHandler` only has references, it can be cheaply recreated on restart. Then `drop_runtime` wouldn't be necessary, IIUC.

Reply:
Mh, good point. After reexamining my counterarguments I have to admit that none of them really holds up. Now moved the creation of `InvoicePayer`, `EventHandler`, and `Router` (for the time being) into the runtime setup, and hence could drop the explicit `{set, drop}` methods here.